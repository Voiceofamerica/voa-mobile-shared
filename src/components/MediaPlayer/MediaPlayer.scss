
@import "../../commonStyles/index.scss";

.mediaPlayer {
  display: flex;
  position: relative;
  flex-direction: column;
  align-content: stretch;
  align-items: stretch;
  flex: 1;

  .loadingBox {
    display: flex;
    justify-content: center;
    align-items: center;
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    background: $voa-light-grey;
    color: black;
    font-size: 40px;
    opacity: 0.8;
  }

  .videoElement {
    flex: 1;
  }
}
