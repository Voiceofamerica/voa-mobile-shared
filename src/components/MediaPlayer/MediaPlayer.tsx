
import * as React from 'react'

export interface Props {
  src: string
  className?: string
  onTogglePlay?: (playing: boolean) => void
  style?: React.CSSProperties
  onLoadDone?: () => void
  playbackRate?: number
  autoPlay?: boolean
  controls?: boolean
}

class MediaPlayer extends React.Component<Props> {
  player: HTMLVideoElement

  componentWillUpdate (nextProps: Props) {
    if (this.player) {
      const { playbackRate = 1 } = nextProps
      this.player.playbackRate = playbackRate
      this.player.defaultPlaybackRate = playbackRate
    }
  }

  togglePlay (play: boolean = this.player.paused) {
    if (play) {
      this.player.play().catch(console.error)
    } else {
      this.player.pause()
    }
  }

  render () {
    const {
      className = '',
      style,
      src,
      autoPlay,
      controls,
    } = this.props

    return (
      <video
        className={className}
        ref={this.setPlayer}
        controls={controls}
        src={src}
        autoPlay={autoPlay}
        onPlay={() => this.triggerTogglePlay(true)}
        onPause={() => this.triggerTogglePlay(false)}
        style={style}
      />
    )
  }

  private triggerTogglePlay = (playing: boolean) => {
    const { onTogglePlay = () => null } = this.props

    onTogglePlay(playing)
  }

  private setPlayer = (player: HTMLVideoElement) => {
    const { playbackRate = 1 } = this.props

    this.player = player

    if (player) {
      player.playbackRate = playbackRate
    }
  }
}

export default MediaPlayer
