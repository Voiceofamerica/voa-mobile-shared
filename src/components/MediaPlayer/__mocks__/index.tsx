
import * as React from 'react'

export default (props) => (
  React.createElement('media-player', { props }, props.children)
)
