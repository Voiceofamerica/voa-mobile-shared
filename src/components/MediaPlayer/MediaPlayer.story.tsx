
import * as React from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'
import { withKnobs, text, boolean, number } from '@storybook/addon-knobs'

import MediaPlayer from './MediaPlayer'

storiesOf('MediaPlayer', module)
  .addDecorator(withKnobs)
  .add('display video', () => (
    <MediaPlayer src='https://av.voanews.com/Videoroot/Pangeavideo/2017/12/1/14/146a3e25-ebcf-444e-b1dd-133cf64a46e4_mobile.mp4' controls playbackRate={number('playbackRate', 1, { min: 0.5, max: 2, step: 0.05, range: true })} />
  ))
  .add('play audio', () => (
    <MediaPlayer src='https://av.voanews.com/clips/VCH/2017/12/06/39923226-d756-47fd-bdf4-b5b6523d9c1f.mp3' controls playbackRate={number('playbackRate', 1, { min: 0.5, max: 2, step: 0.05, range: true })} />
  ))
