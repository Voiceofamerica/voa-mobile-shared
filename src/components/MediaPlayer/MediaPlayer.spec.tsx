
import * as React from 'react'
import { create } from 'react-test-renderer'
import { shallow } from 'enzyme'

import MediaPlayer, { Props } from './MediaPlayer'

jest.mock('../ResilientImage')

describe('<MediaPlayer />', () => {
  describe('shapshots', () => {
    it('should render with the given src', () => {
      const element = create((
        <MediaPlayer src='mySrc' />
      ))

      expect(element).toMatchSnapshot()
    })

    it('should render with the given className', () => {
      const element = create((
        <MediaPlayer src='mySrc' className='myClass' />
      ))

      expect(element).toMatchSnapshot()
    })

    it('should render with the given style', () => {
      const element = create((
        <MediaPlayer src='mySrc' style={{ top: 0, left: 0, right: 0, bottom: 0 }} />
      ))

      expect(element).toMatchSnapshot()
    })

    it('should render with autoplay if prop is set to true', () => {
      const element = create((
        <MediaPlayer src='mySrc' autoPlay />
      ))

      expect(element).toMatchSnapshot()
    })

    it('should render with controls if prop is set to true', () => {
      const element = create((
        <MediaPlayer src='mySrc' controls />
      ))

      expect(element).toMatchSnapshot()
    })

    it('should render with an audio element if audio prop is set to true', () => {
      const element = create((
        <MediaPlayer src='mySrc' audio />
      ))

      expect(element).toMatchSnapshot()
    })
  })

  describe('onTogglePlay', () => {
    it('should be called with true whenever the video starts playing', () => {
      const expected = true
      const onTogglePlay = jest.fn()

      const item = shallow((
        <MediaPlayer src='mySrc' onTogglePlay={onTogglePlay} />
      ))

      item.find('video').simulate('playing')

      expect(onTogglePlay).toHaveBeenCalledTimes(1)
      expect(onTogglePlay).toHaveBeenCalledWith(expected)
    })

    it('should be called with false whenever the video is paused', () => {
      const expected = false
      const onTogglePlay = jest.fn()

      const item = shallow((
        <MediaPlayer src='mySrc' onTogglePlay={onTogglePlay} />
      ))

      item.find('video').simulate('pause')

      expect(onTogglePlay).toHaveBeenCalledTimes(1)
      expect(onTogglePlay).toHaveBeenCalledWith(expected)
    })

    it('should not throw an error if not provided when the video is played or paused', () => {
      const item = shallow((
        <MediaPlayer src='mySrc' />
      ))

      item.find('video').simulate('play')
      item.find('video').simulate('pause')
    })

    it('should be called each time video is played or paused', () => {
      const playCount = 3
      const pauseCount = 2
      const expected = playCount + pauseCount
      const onTogglePlay = jest.fn()

      const item = shallow((
        <MediaPlayer src='mySrc' onTogglePlay={onTogglePlay} />
      ))

      const video = item.find('video')

      for (let i = 0; i < playCount; i++) {
        video.simulate('playing')
      }
      for (let i = 0; i < pauseCount; i++) {
        video.simulate('pause')
      }

      expect(onTogglePlay).toHaveBeenCalledTimes(expected)
    })
  })

  describe('playbackRate', () => {
    it('should update the embedded video playbackRate whenever the prop is changed', () => {
      const expected = 1.5

      const item = shallow<Props>((
        <MediaPlayer src='mySrc' />
      ))

      const mediaPlayer = item.instance() as MediaPlayer
      mediaPlayer.player = {
        playbackRate: 1,
      } as any

      item.setProps({ playbackRate: expected })

      expect(mediaPlayer.player.playbackRate).toEqual(expected)
    })
  })

  describe('togglePlay', () => {
    it('should call play on the player if true is passed in', () => {
      const play = jest.fn(() => Promise.resolve())
      const pause = jest.fn(() => null)

      const item = shallow<Props>((
        <MediaPlayer src='mySrc' />
      ))

      const mediaPlayer = item.instance() as MediaPlayer
      mediaPlayer.player = {
        play,
        pause,
      } as any

      mediaPlayer.togglePlay(true)

      expect(play).toHaveBeenCalledTimes(1)
      expect(pause).toHaveBeenCalledTimes(0)
    })

    it('should call pause on the player if false is passed in', () => {
      const play = jest.fn(() => Promise.resolve())
      const pause = jest.fn(() => null)

      const item = shallow<Props>((
        <MediaPlayer src='mySrc' />
      ))

      const mediaPlayer = item.instance() as MediaPlayer
      mediaPlayer.player = {
        play,
        pause,
      } as any

      mediaPlayer.togglePlay(false)

      expect(play).toHaveBeenCalledTimes(0)
      expect(pause).toHaveBeenCalledTimes(1)
    })

    it('should call play on the player if player is paused and nothing passed in', () => {
      const play = jest.fn(() => Promise.resolve())
      const pause = jest.fn(() => null)

      const item = shallow<Props>((
        <MediaPlayer src='mySrc' />
      ))

      const mediaPlayer = item.instance() as MediaPlayer
      mediaPlayer.player = {
        paused: true,
        play,
        pause,
      } as any

      mediaPlayer.togglePlay()

      expect(play).toHaveBeenCalledTimes(1)
      expect(pause).toHaveBeenCalledTimes(0)
    })

    it('should call pause on the player if player is not paused and nothing passed in', () => {
      const play = jest.fn(() => Promise.resolve())
      const pause = jest.fn(() => null)

      const item = shallow<Props>((
        <MediaPlayer src='mySrc' />
      ))

      const mediaPlayer = item.instance() as MediaPlayer
      mediaPlayer.player = {
        paused: false,
        play,
        pause,
      } as any

      mediaPlayer.togglePlay()

      expect(play).toHaveBeenCalledTimes(0)
      expect(pause).toHaveBeenCalledTimes(1)
    })
  })
})
