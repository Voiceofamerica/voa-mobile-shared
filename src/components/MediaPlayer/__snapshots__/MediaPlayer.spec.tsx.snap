// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MediaPlayer /> shapshots should render with autoplay if prop is set to true 1`] = `
<video
  autoPlay={true}
  className=""
  controls={undefined}
  onPause={[Function]}
  onPlay={[Function]}
  src="mySrc"
  style={undefined}
/>
`;

exports[`<MediaPlayer /> shapshots should render with controls if prop is set to true 1`] = `
<video
  autoPlay={undefined}
  className=""
  controls={true}
  onPause={[Function]}
  onPlay={[Function]}
  src="mySrc"
  style={undefined}
/>
`;

exports[`<MediaPlayer /> shapshots should render with the given className 1`] = `
<video
  autoPlay={undefined}
  className="myClass"
  controls={undefined}
  onPause={[Function]}
  onPlay={[Function]}
  src="mySrc"
  style={undefined}
/>
`;

exports[`<MediaPlayer /> shapshots should render with the given src 1`] = `
<video
  autoPlay={undefined}
  className=""
  controls={undefined}
  onPause={[Function]}
  onPlay={[Function]}
  src="mySrc"
  style={undefined}
/>
`;

exports[`<MediaPlayer /> shapshots should render with the given style 1`] = `
<video
  autoPlay={undefined}
  className=""
  controls={undefined}
  onPause={[Function]}
  onPlay={[Function]}
  src="mySrc"
  style={
    Object {
      "bottom": 0,
      "left": 0,
      "right": 0,
      "top": 0,
    }
  }
/>
`;
